cmd/mdatagen/metadata-schema.yaml

```diff
@@ -78,6 +78,8 @@ attributes:
     enum: [string]
     # Required: attribute value type.
     type: <string|int|double|bool|bytes|slice|map>
+    # Optional: indicates whether the attribute is optional or not.
+    optional:
```

Comment:
Would it make more sense to indicate whether an attribute is optional or required in the metric/event's schema, rather than at the global attributes definition level? We can then handle use cases such as attribute X is required for metric A but optional for metric B.

Reply:
Wdyt about a mix of the two approaches -

```
metrics:
  <metric-name>:
    attributes: ["attribute-name-1", "attribute-name-2"]
    attribute_settings:
      attribute-name-1:
        optional: true
```

Instead of the optional_attributes array, we have a map for settings which we can add to.

Approach 1 would be more clean, but yeah, i am not sure if we want to take on breaking config changes.

Reply:
We have [a long-standing issue](https://github.com/open-telemetry/opentelemetry-collector/issues/10726) that still needs to be implemented that would benefit from swicthing from the attributes list to the map. 

@sincejune I think it's fine to go with a breaking change it's a separate command line executable that independent form the collector core

Reply:
I feel that moving the entire attribute definitions to the metrics/events level might result in too many duplicated attributes. Are we okay with this approach? Perhaps allowing additional/override configurations at the metrics/events level would be a better solution here?

Reply:
Updated the PR.